Google Books Implementation

LilAttacker opened the pull request:
# Added Dynamic Search Capability For Google Books.

## Description

Prior, it was all hard coded and was a very basic implementation. The code was janky and used a global class variable to hold it all. The new implementation now uses parameter arguments and dynamically performs the search itself.

Not only that, it formats the query automatically so you just need to define some variables. It also has some error checking to make sure enough variables were given to make a query and that improper arguments are not used.

Before, all data was stored in a json. Now all book data is stored in a dataclass which is placed in a list and returned as a json. This makes it more malleable and easier to adjust if we want to in the future.

Most of the implementation is split over 4 functions.
- google_books_search: The one you call and that. Checks if proper arguments were given and calls the other functions. Also makes the call to Google Books API itself when the query is ready.
- query_assembler: Creates the search query for Google Books. This one calls parameter_handler.
- parameter_handler: Formats the arguments to be readable by Google Books. Only used for when searching for specific things like title or author.
- attr_collecter: Collects appropriate attributes from each book if it can be found and saves it all in the BookAttr() dataclass. The only books it does not save are ones that lack the "ISBN_13" industry identifier.

I also added all necessary test cases for all functions already.

I believe this is practically as close to final implementation as we can get at the moment.

## Type of Change

- [ ] Bugfix
- [X] New Feature
- [X] Documentation Update
- [X] Other (please describe): Test Cases and Greater Control

## Checklist

- [X] I have read the contributing guidelines.
- [X] I have followed the code style of this project.
- [X] I have added tests to cover my changes.
- [X] I have updated the documentation (if applicable).
- [X] My changes do not generate new warnings.

## Related Issue

#38
#19
#33 

## Additional Notes

Removed Some Files
There were some unused backend files that were either for functionality we decided to drop or for recording some outputs to view and test against. I removed some and moved another to the tests folder to be used during pytest.

Updated Main
After some testing, I managed to find that I no longer needed the pytest vs python separate import calls and could safely remove them. This is pretty minor but was so satisfying to do.

ederic-oytas reviewed (commented):
Daniel, I've reviewed your PR and made a couple of comments. My biggest comments:
- The logic of `parameter_handler` and `query_assembler` can simplified into a single function.
- The logic of handling API parameters should be placed in `api/books.py` instead of here. The functionality of calling the Google Books API should be placed here as its own function, not an endpoint.

David1000734 commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+    """
+
+    # Assembles the query here.
+    complete = (
```
I'm a little confused with what this part here is doing. I see that it should assemble a query here but I'm not familiar with the syntax that follows below. Can we get some comments that explains how it assembles the query using the inline if statements?

LilAttacker replied:
It was just using some concatenation and ternary statements, that was all. I added some more comments still.

David1000734 commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+
+    # Make call to Google Books with assembled query.
+    # Save only the books themselves if any were found.
+    response = requests.get(api_http + query + api_key + api_max)
```
I think we should use this method as well. It would significantly improve readability and make it easier to do other calls in the future.

LilAttacker replied:
I'll make the change.

David1000734 commented on server/tests/test_api.py:
```diff
@@ -135,3 +136,231 @@ def test_predict_examples(self, client):
+            test_output = json.load(f)
+
+        # Setup Mocker
+        with requests_mock.Mocker() as m:
```
Good job with the tests :D

LilAttacker replied:
Thank you

David1000734 reviewed (approved):
Added some comments. Overall, very well done. Just some things to consider for the future.

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+    # Basic aspects of the Google Books API call.
+    api_http = "https://www.googleapis.com/books/v1/volumes?q="
+    api_key = "&key=" + str(os.getenv("API_KEY"))
+    api_max = "&maxResults=40"
```
Fair. I was originally told that this value should never change, but making it more malleable is always good.

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+
+# Dataclass to hold book attributes from the google books api call.
+@dataclass(frozen=True, kw_only=True)
+class BookAttr:
```
Sure. I'll change that.

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+
+    # Basic aspects of the Google Books API call.
+    api_http = "https://www.googleapis.com/books/v1/volumes?q="
+    api_key = "&key=" + str(os.getenv("API_KEY"))
```
I'm aware that it returns `None` if no key is found, that's preferable in my opinion.

Using the method you defined would actually get a `KeyError` if the key wasn't found which is what occurs when this is uploaded to the GitHub.

LilAttacker commented on server/tests/test_api.py:
```diff
@@ -135,3 +136,231 @@ def test_predict_examples(self, client):
+            assert result.status_code == HTTPStatus.OK
+
+
+def test_parameter_handler():
```
But this is? Did you mean you want this in another file in `tests/`?

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+    Returns:
+        complete: Fully assembled and formatted Google Books API search
+                  query.
+    """
```
I can resolve that.

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+    if not phr and complete:
+        complete = complete[1:]
+
+    return complete
```
I'll combine `parameter_handler` and `query_assembler` but the method you used wouldn't work.

It would incorrectly add additional phrases and strings when it was unnecessary. I messed around with it a little but ultimately found using string concatenation was the most reliable still.

LilAttacker commented on server/tabby_server/__main__.py:
```diff
@@ -1,21 +1,31 @@
-app.register_blueprint(books_test, url_prefix="/test")
+# Test Python Files.
+# YOLO or Image Recognition
+app.register_blueprint(yolo_test.yolo_test, url_prefix="/yolo")
```
I can add an example of the routing, but otherwise I'm not entirely sure what the issue is.

LilAttacker commented on server/tabby_server/services/library.py:
```diff
@@ -0,0 +1,231 @@
+            return param
+
+
+def query_assembler(phr="", tit="", aut="", pub="", sub="", isb=""):
```
Changed.

LilAttacker commented:
Alright so I changed some things around.

- Added example of route to function in __main__.py.
- Made the max result variable and api key as global variables.
- Added an excerpt variable to the Book() dataclass. This takes the first 50ish characters of summary after initialization.
- Whole bunch of formatting and reconstruction of the code. Combined query_assembler and parameter_handler into one function. Streamlined some code too make it easier to parse.
- Changed output to just return one book instead of a list.
- Added more comments to better describe things.
- Changes tests to work and properly test all new functionality.

ederic-oytas commented on server/tabby_server/services/library.py:
```diff
@@ -15,151 +16,144 @@
+        if len(self.summary) > 50:
+            object.__setattr__(self, "excerpt", f"{self.summary[:46]}...")
+        else:
+            object.__setattr__(self, "excerpt", self.summary)
```
I think it's better to make this a property instead of an actual attribute:
```python
@property
def excerpt(self) -> str:
    return self.summary[:46]
```
That way, it's not modifiable; we know it's a calculated property.

ederic-oytas replied:
Also, indices don't need to be in bounds when taking a slice.

ederic-oytas commented on server/tabby_server/services/library.py:
```diff
@@ -15,151 +16,144 @@
+        + (f"+inauthor:{author}" if author else "")
+        + (f"+inpublisher:{publisher}" if publisher else "")
+        + (f"+subject:{subject}" if subject else "")
+        + (f"+isbn:{isbn}" if isbn else "")
```
This looks good.

ederic-oytas commented on server/tabby_server/services/library.py:
```diff
@@ -15,151 +16,144 @@
 
-
-def attr_collecter(book={}):
+def attr_collecter(book: dict = {}) -> dict[str, str]:
```
I think this is better named `collect_attrs`

ederic-oytas reviewed (approved):
I think this looks good. See my comments listed. I think we'll just merge these changes then fix some issues if they appear in a future PR.
